### PR TITLE
Fix glb/gltf export: texture not tiled / invalid format for pictures

### PR DESCRIPTION
**Pull request by Popov72**
See https://github.com/BabylonJS/Babylon.js/issues/8514

### CODE REVIEW

File: serializers/src/glTF/2.0/glTFMaterialExporter.ts [hunk 1]
```diff
@@ -515,7 +515,15 @@ export class _GLTFMaterialExporter {
...
-                const canvas = engine.getRenderingCanvas();
+                const canvas0 = engine.getRenderingCanvas();
+
+                let canvas: Nullable<HTMLCanvasElement> = document.createElement("canvas");
+
+                canvas.width = canvas0?.width ?? 0;
+                canvas.height = canvas0?.height ?? 0;
+
+                var destCtx = canvas.getContext('2d');
+                destCtx!.drawImage(canvas0!, 0, 0);
```
Comment (Popov72) [inline serializers/src/glTF/2.0/glTFMaterialExporter.ts:526]: I think the block of code above that copies the canvas is needed, else it's possible we render a picture whereas the previous one is still not processed by the code below (which converts the canvas to a blob then to a base64 string), overwritting the previous data in the canvas.

Or are we sure all the promises involved in the process will be completed right away, so in effect all the conversion code is synchronous (in which case no need to copy the canvas)? In effect, it's only `fileReader.readAsDataURL` that will trigger `fileReader.onLoad` that is/can be asynchronous. If we are sure it won't be done asynchronously, I can remove my code.

I tried to have a look at the [MDN](https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsDataURL) and the [spec](https://w3c.github.io/FileAPI/#readAsDataURL), but I can't see anything saying that `readAsDataURL` would be synchronous...

**Comment by Popov72** [comment]
A bit of explanation regarding the "invalid format" for pictures:

The problem was that the JSON `images` node is filled by looping through `this.images` (line 953 in `glTFExporter.ts`) whereas the binary data of the pictures were written to the glb file in the order of the key of `this._imageData` because the loop was `for (key in this._imageData)`.

I think that by default (at least in Chrome - I don't think it is a requirement of the spec) the keys are looped in the order they were created (else I think we would have seen the problem earlier). So, if insertions are made in `_images` in the same order than in `_imageData` then everything is ok. However, because of asynchronous operations, it's possible we perform the insertions out of order, meaning `_images[0]` does not correspond to the first object inserted in `_imageData`.

This one made me sweat quite a bit, as the bug did not appear all the time because of the asynchronous dimension and because for a long time I was looking at the wrong place (as some pictures were invalid, I was looking for problems in the texture management/conversion code...).

File: serializers/src/glTF/2.0/glTFMaterialExporter.ts (hunk 1 above)
Comment (deltakosh) [reply serializers/src/glTF/2.0/glTFMaterialExporter.ts:526]: I think this is fine to be secure

**Comment by deltakosh** [comment]
Good job..this kind of race condition are terrible to fix
Thanks!

**Pull request merged**